man/man3/Media.3
.TH "Media" 3 "Fri Oct 7 2011" "Version 1.0" "Launchpad" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Media \- 
.SH SYNOPSIS
.br
.PP
.PP
Inherits \fBDatabaseObject\fP.
.PP
Inherited by \fBImage\fP.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBattach_file\fP ($file)"
.br
.ti -1c
.RI "\fBsave\fP ()"
.br
.ti -1c
.RI "\fBdestroy\fP ()"
.br
.ti -1c
.RI "\fBforce_mimetype\fP ()"
.br
.ti -1c
.RI "\fBauthor\fP ()"
.br
.ti -1c
.RI "\fBfile_path\fP ()"
.br
.ti -1c
.RI "\fBname_of_file\fP ()"
.br
.ti -1c
.RI "\fBsimple_type\fP ()"
.br
.ti -1c
.RI "\fBsize_as_text\fP ()"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBextension\fP ($filename)"
.br
.ti -1c
.RI "static \fBfind_last_created\fP ($limit=1)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fB$id\fP"
.br
.ti -1c
.RI "\fB$filename\fP"
.br
.ti -1c
.RI "\fB$type\fP"
.br
.ti -1c
.RI "\fB$size\fP"
.br
.ti -1c
.RI "\fB$caption\fP"
.br
.ti -1c
.RI "\fB$description\fP"
.br
.ti -1c
.RI "\fB$author_id\fP"
.br
.ti -1c
.RI "\fB$errors\fP = array()"
.br
.ti -1c
.RI "\fB$upload_errors\fP"
.br
.in -1c
.SS "Static Protected Attributes"

.in +1c
.ti -1c
.RI "static \fB$table_name\fP = 'media'"
.br
.ti -1c
.RI "static \fB$db_fields\fP = array( 'id', 'filename', 'type', 'size', 'caption', 'description', 'author_id')"
.br
.in -1c
.SH "Member Function Documentation"
.PP 
.SS "Media::attach_file ($file)"Sets \fBImage\fP attributes. Useage: attach_file( $_FILE['uploaded_file'] );
.PP
\fBParameters:\fP
.RS 4
\fI$file\fP - An associative array of items uploaded to the current script via the HTTP POST method 
.RE
.PP
\fBReturns:\fP
.RS 4
boolean 
.RE
.PP

.PP
Reimplemented in \fBImage\fP.
.SS "Media::author ()"Reurns author of media by author_id 
.PP
\fBReturns:\fP
.RS 4
\fBUser\fP 
.RE
.PP

.SS "Media::destroy ()"Removes DB entry and file on server. 
.PP
\fBReturns:\fP
.RS 4
boolean 
.RE
.PP

.SS "static Media::extension ($filename)\fC [static]\fP"Returns returns the extension type of media. 
.PP
\fBReturns:\fP
.RS 4
string 
.RE
.PP

.SS "Media::file_path ()"Returns the file path of the media file. 
.PP
\fBReturns:\fP
.RS 4
string 
.RE
.PP

.SS "static Media::find_last_created ($limit = \fC1\fP)\fC [static]\fP"Returns last created media 
.PP
\fBParameters:\fP
.RS 4
\fI$limit\fP = 1 - Specifies the amount of media files to return 
.RE
.PP
\fBReturns:\fP
.RS 4
\fBMedia\fP 
.RE
.PP

.SS "Media::force_mimetype ()"Guesses the content type of a file by looking for certain magic byte sequences at specific positions within the file. If Magic File isn't available mime is guessed based on the file extention
.PP
\fBReturns:\fP
.RS 4
string 
.RE
.PP

.SS "Media::name_of_file ()"Returns the base name of the file without directories. 
.PP
\fBReturns:\fP
.RS 4
string 
.RE
.PP

.SS "Media::save ()"Saves image 
.PP
\fBReturns:\fP
.RS 4
boolean 
.RE
.PP

.PP
Reimplemented from \fBDatabaseObject\fP.
.SS "Media::simple_type ()"Returns the type of media 
.PP
\fBReturns:\fP
.RS 4
string eg: image, video, audio, text 
.RE
.PP

.SS "Media::size_as_text ()"Returns the size of the file as string. 
.PP
\fBReturns:\fP
.RS 4
string 
.RE
.PP

.SH "Member Data Documentation"
.PP 
.SS "Media::$upload_errors"\fBInitial value:\fP
.PP
.nf
 array(
                UPLOAD_ERR_OK                   => 'No errors',
                UPLOAD_ERR_INI_SIZE     => 'Larger than upload_max_filesize',
                UPLOAD_ERR_FORM_SIZE    => 'Larger than MAX_FILE_SIZE',
                UPLOAD_ERR_PARTIAL              => 'Partial upload',
                UPLOAD_ERR_NO_FILE              => 'No file',
                UPLOAD_ERR_NO_TMP_DIR   => 'No temporary directory',
                UPLOAD_ERR_CANT_WRITE   => 'Can't write to disk',
                UPLOAD_ERR_EXTENSION    => 'File upload stopped by extension'
        )
.fi


.SH "Author"
.PP 
Generated automatically by Doxygen for Launchpad from the source code.


lib/media.class.php
<?php

/**
 * Base class for uploading and modifying of media files
 *
 * @author Jonah Werre <jonahwerre@gmail.com>
 * @version 1.0
 * @copyright Jonah Werre <jonahwerre@gmail.com>, 28 June, 2011
 * @package DatabaseObject
 **/
class Media extends DatabaseObject
{
	/**
	 * The name of the database table
	 * @var string
	 **/
	protected static $table_name = "media";
	/**
	 * The name of the database fields in $table_name
	 * @var array
	 **/
	protected static $db_fields = array( 'id', 'filename', 'type', 'size', 'caption', 'description', 'author_id');
	/**
	 * temporary path the the file
	 * @var string
	 **/
	private $temp_path;
	/**
	 * The unique id of the image
	 * @var string
	 **/
	public $id;
	/**
	 * The media full filename
	 * @var string
	 **/
	public $filename;
	/**
	 * The mime type of media 
	 * @var string
	 **/
	public $type;
	/**
	 * The size of the media file
	 * @var string
	 **/
	public $size;
	/**
	 * A caption for the media file
	 * @var string
	 **/
	public $caption;
	/**
	 * A description for the media file
	 * @var string
	 **/
	public $description;
    /**
     * The uniqie id of the User who created the Media
     * @var intiger
     **/
	public $author_id;
	/**
	 * A collection of errors created durring Media uploading
	 * @var string
	 **/
	public $errors = array();
	/**
	 * A collection of predefined errors
	 * @var string
	 **/
	public $upload_errors = array(
		UPLOAD_ERR_OK 			=> "No errors",
		UPLOAD_ERR_INI_SIZE  	=> "Larger than upload_max_filesize",
		UPLOAD_ERR_FORM_SIZE 	=> "Larger than MAX_FILE_SIZE",
		UPLOAD_ERR_PARTIAL 		=> "Partial upload",
		UPLOAD_ERR_NO_FILE 		=> "No file",
		UPLOAD_ERR_NO_TMP_DIR 	=> "No temporary directory",
		UPLOAD_ERR_CANT_WRITE 	=> "Can't write to disk",
		UPLOAD_ERR_EXTENSION 	=> "File upload stopped by extension"
	);
		
	/**
	 * Sets Image attributes.
	 * Useage: attach_file( $_FILE['uploaded_file'] );
	 *
	 * @param array $file - An associative array of items uploaded to the current script via the HTTP POST method
	 * @return boolean
	 */
	public function attach_file($file)
	{
		if( !$file || empty($file) || !is_array($file) ){
			array_push( $this->errors, "No file was uploaded" );
			return false;
		}elseif($file['error'] != 0){
			array_push( $this->errors, $this->upload_errors[$file['error']] );
			return false;
		}else{
			$this->temp_path = $file['tmp_name'];
			$this->filename = UPLOADS_URL."/".strtolower( str_replace(" ","_", basename($file['name']) ) );
			$this->type = $file['type'];
			$this->size = $file['size'];
			return true;
		}
	}
	
	/**
	* Saves image
	* @return boolean
	*/
	public function save()
	{
		if( isset($this->id) ){
			return $this->update();
		}else{
			if(!empty($this->errors)) { return false; }
			// check and see if filename and temp_path exists
			if(empty($this->filename) || empty($this->temp_path)) {
				array_push( $this->errors, "The file location was not available.");
				return false; 
			}
			
			$target_path = $this->file_path();
			// make sure the file exists
			if(file_exists($target_path)){
				array_push( $this->errors, "The file {$this->name_of_file()} already exists");
				return false;
			}
			// move the file to permanent location
			if ( move_uploaded_file($this->temp_path, $target_path) ) {
				if($this->create()){
					unset($this->temp_path);
					return true;
				}
			} else {
				array_push( $this->errors, "file ulpoad failed");
				return false;
			}
		}
	}
	
	
	/**
	* Removes DB entry and file on server.
	* @return boolean
	*/
	public function destroy() {
		$this->errors = array();
		if($this->delete()) {
			
			$target_path = $this->file_path();
			
			if (file_exists($target_path)) {
				
				if (unlink($target_path)) {
					array_push( $this->errors, "The file at $target_path was deleted");
					return true;
				}else{
					array_push( $this->errors, "Could not delete file at $target_path");
					return false;
				}
				
			}else{
				array_push( $this->errors, "Could not find the file at $target_path");
				return false;
			}
		
		}else{
			array_push( $this->errors, "No file reference for $target_path. No Files were deleted");
			return false;
		}
		
	}
	
	/**
	 * Guesses the content type of a file by looking for certain magic byte sequences at specific positions within the file.
	 * If Magic File isn't available mime is guessed based on the file extention
	 *
	 * @return string 
	 **/
	public function force_mimetype()
	{
		if(function_exists('finfo_open')){
			$file = finfo_open(FILEINFO_MIME_TYPE);
			$info = finfo_file($file, $this->file_path() );
			$this->type = $info;
			finfo_close($file);		
		}else{
			$mime_types = array(
				'txt' => 'text/plain', 'htm' => 'text/html', 'html' => 'text/html', 'php' => 'text/html', 'css' => 'text/css', 'js' => 'application/javascript', 'json' => 'application/json', 'xml' => 'application/xml', 'swf' => 'application/x-shockwave-flash', 'flv' => 'video/x-flv',
				// images
				'png' => 'image/png', 'jpe' => 'image/jpeg', 'jpeg' => 'image/jpeg', 'jpg' => 'image/jpeg', 'gif' => 'image/gif', 'bmp' => 'image/bmp', 'ico' => 'image/vnd.microsoft.icon', 'tiff' => 'image/tiff', 'tif' => 'image/tiff', 'svg' => 'image/svg+xml', 'svgz' => 'image/svg+xml',
				// archives
				'zip' => 'application/zip', 'rar' => 'application/x-rar-compressed', 'exe' => 'application/x-msdownload', 'msi' => 'application/x-msdownload', 'cab' => 'application/vnd.ms-cab-compressed',
				// audio/video
				'mp3' => 'audio/mpeg', 'qt' => 'video/quicktime', 'mov' => 'video/quicktime','mp4' => 'video/quicktime','m4v' => 'video/quicktime','ogg' => 'audio/ogg', 'oga' => 'audio/ogg','ogv' => 'video/ogg',
				// adobe
				'pdf' => 'application/pdf', 'psd' => 'image/vnd.adobe.photoshop', 'ai' => 'application/postscript', 'eps' => 'application/postscript', 'ps' => 'application/postscript',
				// ms office
				'doc' => 'application/msword', 'rtf' => 'application/rtf', 'xls' => 'application/vnd.ms-excel', 'ppt' => 'application/vnd.ms-powerpoint',
				// open office
				'odt' => 'application/vnd.oasis.opendocument.text', 'ods' => 'application/vnd.oasis.opendocument.spreadsheet',
			);
			$ext = strtolower( pathinfo($this->filename, PATHINFO_EXTENSION) );
			if (array_key_exists($ext, $mime_types)) {
				$this->type = $mime_types[$ext];
			} else {
				$this->type = 'application/octet-stream';
			}
		}
		
	}
	
	/**
	* Reurns author of media by author_id
	* @return User
	**/
	public function author()
	{
		$user = User::find_by_id( $this->author_id );
		return !empty($user) ? $user : false;
	}
	
	/**
	* Returns the file path of the media file.
	* @return string
	*/
	public function file_path()
	{
		$path = substr( $this->filename, strlen(BASE_URL) );
		$path = str_replace('/', DS, SITE_ROOT.'/'.$path);
		return $path;  
	}	
	/**
	* Returns the base name of the file without directories.
	* @return string
	*/
	public function name_of_file()
	{
		$path = substr( $this->filename, strlen(UPLOADS_URL) );
		return ltrim($path, '/');  
	}	
    /**
     * Returns the type of media
     * @return string eg: image, video, audio, text
     **/
    public function simple_type()
    {
        return strstr($this->type, '/', true);
    }
	
	/**
	* Returns returns the extension type of media.
	* @return string
	*/
	public static function extension($filename) {
		//$result_string = substr($this->type,0,strpos($this->type,"/")+strlen("/")-1); 
		return strtolower( pathinfo($filename, PATHINFO_EXTENSION) );
	}
	
	/**
	* Returns the size of the file as string.
	* @return string
	*/
	public function size_as_text() {
		if($this->size < 1024) {
			return "{$this->size} bytes";
		} elseif($this->size < 1048576) {
			$size_kb = round($this->size/1024);
			return "{$size_kb} KB";
		} else {
			$size_mb = round($this->size/1048576, 1);
			return "{$size_mb} MB";
		}
	}
    
    /**
    * Returns last created media
	* @param integer $limit = 1 - Specifies the amount of media files to return
    * @return Media
    **/
    public static function find_last_created($limit = 1)
    {
        $sql = "SELECT * FROM " . self::$table_name; 
        $sql .= " ORDER BY created DESC LIMIT ".$limit;
        $result = static::find_by_sql($sql);
        return !empty($result) ? $result : false;
    }
	
}

?>
